### DISCUSSION TITLE
[React]  virtual Dom에 대해 설명해보세요.

### DISCUSSION THREAD

**Discussion by bona373737** [wanted-9th-3team/frontend-study#7] [프론트 기술 질문 공간]
리액트의 동작원리를 이해하고 있는지를 평가하는 질문입니다.

**Comment by WooGie911** [comment]
DOM이란 웹페이지에 들어가있는 html element들을 트리의 형태의 구조로 표현한 것입니다. 
리액트의 가상돔은 실제 돔과 같은 내용을 담고있는 복사본이고 이 복사본은 자바스크립트 객체 형태로 메모리상에 저장되어 있습니다. 
리액트는 항상 두개의 가상돔을 갖고있는데 첫번째는 변경 이전의 내용을 담고있고, 두번째는 변경 이후 보여질 내용을 담고 있습니다. 
실제 돔이 변경되기 이전에 리액트는 두개의 가상돔을 비교해서 정확히 어떤 부분이 바뀌었는지 효율적으로 비교하여 파악합니다. 
이 과정을 Diffing이라고 합니다. 
diffing을 통해서 변경된 부분을 파악한 후, Batch update를 수행하여 실제 돔에 한번에 적용시켜 줍니다. 
그 과정을 Resconsiliation 재조정 이라고 합니다. 
돔을 변경 해야 하는 경우 구성요소에 따라 여러번 렌더링이 되어야 하는 문제가 있는데 
가상돔을 적용하면 한번 렌더링이 일어나면 된다는 장점이 있습니다.

**Comment by michoball** [comment]
JS에서 DOM요소를 조작하면  JS에 구현한 동작에 따라 DOM요소들이 바뀌고 그에 맞게 다른 요소들도 함께 재배치되게 됩니다.

여기서 가상돔 ( Virtual DOM )은 실제 DOM 요소를 쓰지않고 DOM을 닮은 JS 객체를 이용해서 실제 DOM 요소가 어떻게 바뀔지를 미리 체크헤

 바뀌는 부분만 실제 DOM에 반영하여 속도를 향상하고 유저에서 좋은 UX를 제공해주는 역할을 합니다.

**Comment by skyhanull** [comment]
react는 뒤에서 실제 DOM을 모두 복사해서 virtual DOM을 만든다.
돔이란 (document object model)의 줄임말로  html요소들을 트리 구조로 만든 것을 의미합니다.
가상돔은 이런 돔을 복사한 일종의 복사본입니다.
만약 리액트를 이용해서 virtual dom의 어떤 노드를 수정한다면 리액트는 새로운 virtual dom(updated virtual dom)을 만들고 이전 virtual dom 트리와 비교하여 만약 일치하지 않는 노드를 발견한다면 리액트는 실제 dom 트리에서 해당 노드만 업데이트 시켜줍니다.

**Comment by BaekYuri** [comment]
React에서는 Virtual DOM이라는 가상의 레이어를 만든 후, 변경 사항이 발생되었을 때 한번에 DOM에 업데이트를 합니다. 그러므로 렌더링을 여러번 할 필요 없이 한 번만 하게 됩니다.
